src/logging/Logger.cpp

```diff
@@ -1,53 +1,158 @@
-#include <boost/log/core.hpp>
-#include <boost/log/expressions.hpp>
+#include <boost/log/attributes/timer.hpp>
+#include <boost/log/sources/severity_feature.hpp>
 #include <boost/log/trivial.hpp>
 #include <boost/log/utility/setup/common_attributes.hpp>
-#include <iosfwd>
 #include <utility>
+#include <utils/assertion.hpp>
+
+#include "logging/LogConfiguration.hpp"
+#include "logging/Logger.hpp"
 
 namespace precice {
 namespace logging {
 
+/// A feature that adds \ref LogLocation info to log attributes
+template <class BaseT>
+class precice_feature : public BaseT {
+public:
+  using char_type       = typename BaseT::char_type;
+  using threading_model = typename BaseT::threading_model;
+
+  precice_feature()                             = default;
+  precice_feature(const precice_feature &other) = default;
+  template <typename ArgsT>
+  precice_feature(ArgsT const &args){};
+
+  using open_record_lock = typename boost::log::strictest_lock<
+      boost::lock_guard<threading_model>,
+      typename BaseT::open_record_lock,
+      typename BaseT::add_attribute_lock,
+      typename BaseT::remove_attribute_lock>::type;
+
+protected:
+  template <typename ArgsT>
+  boost::log::record open_record_unlocked(ArgsT const &args);
+};
+
+/// Provides keywords to pass \ref LogLocation related info to the precice_feature
+/// @see PRECICE_LOG_IMPL
+namespace keywords {
+BOOST_PARAMETER_KEYWORD(line_ns, line);
+BOOST_PARAMETER_KEYWORD(file_ns, file);
+BOOST_PARAMETER_KEYWORD(func_ns, func);
+} // namespace keywords
+
+namespace {
+/// Runs the function F on end of lifetime.
+template <class F>
+struct ScopeExit {
+  F _f;
+  ScopeExit(F f)
+      : _f(f){};
+  ~ScopeExit()
+  {
+    _f();
+  }
+};
+} // namespace
+
+/// Adds log attributes to the current logger based on the \ref LogLocation info
+template <typename BaseT>
+template <typename ArgsT>
+boost::log::record precice_feature<BaseT>::open_record_unlocked(ArgsT const &args)
+{
+  // Extract the named argument from the parameters pack
+  int         line_value = args[keywords::line | -1];
+  std::string file_value = args[keywords::file | std::string()];
+  std::string func_value = args[keywords::func | std::string()];
+
+  PRECICE_ASSERT(line_value >= 0);
+  PRECICE_ASSERT(!file_value.empty());
+  PRECICE_ASSERT(!func_value.empty());
+
+  // Process extracted tags
+  using boost::log::attributes::constant;
+  boost::log::attribute_set &attrs = BaseT::attributes();
+  {
+    auto res = BaseT::add_attribute_unlocked("Line", constant<int>(line_value));
+    PRECICE_ASSERT(res.second);
+  }
+  {
+    auto res = BaseT::add_attribute_unlocked("File", constant<std::string>(file_value));
+    PRECICE_ASSERT(res.second);
+  }
+  {
+    auto res = BaseT::add_attribute_unlocked("Function", constant<std::string>(func_value));
+    PRECICE_ASSERT(res.second);
+  }
+
+  ScopeExit guard([&attrs] {
+    attrs.erase("Line");
+    attrs.erase("File");
+    attrs.erase("Function");
+  });
```

Comment:
Instead of the `ScopeExit` solution here, couldn't we just define a lambda a few lines further up? Would be a bit more readable as the function is closer to the actual call.

Reply:
Not sure what you mean. Could you provide an example?

Reply:
Thinking about it: I mean the scope exit is right after this. Couldn't we just execute the function in place?

Reply:
Actually, we could save the result, cleanup and then return, making this way easier.

Reply:
Done. Thanks for complaining :smiley: